### ISSUE TITLE
Dataset for clustering: discussion

### ISSUE THREAD

**Issue by hainm** [Amber-MD/cpptraj#189]
from this PR: #183 

Dan's comment:

> I'm delaying on this a little bit because I'm wondering if all clustering data shouldn't be stored in either one specialized DataSet or in multiple DataSets with different aspects. Former would allow cluster output to be used with the DataFile framework more easily, latter would be easier to implement in the short run but maybe harder to maintain. Maybe this should be a separate issue? Not sure...

Hai:

I am ok with delaying (and I don't need this feature now). My original aim for this PR is to make cpptraj a bit similar to sklearn. Just give very dummy example here.

``` python
In [3]: from sklearn.cluster import k_means

In [4]: k_means(xyz, n_clusters=10)
Out[4]: 
(array([[ 15.48531055,  28.5936985 ,  16.56096268, ...,  34.20779419,
           9.24066639,  16.65553093],
        [ 15.07740974,  29.1335144 ,  17.32461929, ...,  33.20933914,
           7.98652649,  15.00128174],
        [ 14.73693371,  28.61953735,  16.14261436, ...,  32.55017853,
           9.01970196,  17.33253098],
        ..., 
        [ 15.06157207,  28.82535172,  17.34188652, ...,  34.29132462,
           8.27471733,  16.50368881],
        [ 15.55458927,  28.54844856,  17.18908691, ...,  -0.17751557,
           8.35964298,  18.69264793],
        [ 14.77806568,  29.08259392,  16.52130318, ...,  33.91731262,
           9.42454243,  16.95419884]]),
 array([8, 3, 7, 1, 6, 0, 5, 4, 9, 2], dtype=int32),
 1.3271346688270569e-08)
```

``` python
Returns
-------
centroid : float ndarray with shape (k, n_features)
    Centroids found at the last iteration of k-means.

label : integer ndarray with shape (n_samples,)
    label[i] is the code or index of the centroid the
    i'th observation is closest to.

inertia : float
    The final value of the inertia criterion (sum of squared distances to
    the closest centroid for all observations in the training set).
```

**Comment by drroe** [comment]
> My original aim for this PR is to make cpptraj a bit similar to sklearn. 

I think this is a fine idea. I just wonder which approach I should take to get there. Ideally you'd like access to more than just representative frames, right? Like the actual frame numbers in each cluster, the cluster silhouette data, pseufo-F, DBI, etc.

**Comment by hainm** [comment]
but for my need (from my past work), I think getting representative frame and frame label for each cluster are enough. 

```python
from pytraj.cluster import kmeans

refframe_indices, labels = kmeans(traj, n_clusters=50)
# slicing traj to get coordinates of ref frame
traj[refframe_indices]

# save to pdb
traj[refframe_indices].save('test.pdb')
```

**Comment by hainm** [comment]
>  Like the actual frame numbers in each cluster, the cluster silhouette data, pseufo-F, DBI, etc.

Per " the cluster silhouette data, pseufo-F, DBI, etc.", I have never used them. So I am not sure about how I can use. May be we just dump them to a single String dataset? If anyone want to use, he/she can parse the string.

Per 'actual frame numbers in each cluster`, Python has excellent class, called Counter :D

```python
In [16]: data = kmeans(traj, n_clusters=20)

In [17]: data
Out[17]: array([16, 18, 17, ...,  0,  0,  0], dtype=int32)

In [18]: Counter(data)
Out[18]: 
Counter({0: 20,
         1: 17,
         2: 12,
         3: 12,
         4: 6,
         5: 4,
         6: 4,
         7: 3,
         8: 3,
         9: 3,
         10: 3,
         11: 2,
         12: 2,
         13: 2,
         14: 2,
         15: 2,
         16: 1,
         17: 1,
         18: 1,
         19: 1})
```

**Comment by hainm** [comment]
Close this since Pytraj knows how to get cpptraj stdout now. So just parse the output.
